data/tutorials/language/3ds_07_monads.md

```diff
@@ -0,0 +1,822 @@
+
+```ocaml
+let e x = (x, "")
+```
+
+And now we can write `id' (e 5)` instead of `id' (5, "")`.
+
+**Where's the Monad?** The work we just did was to take functions on integers
+and transform them into functions on integers paired with log messages. We can
+think of these "upgraded" functions as computations that log. They produce
+metaphorical boxes, and those boxes contain function outputs as well as log
+messages.
+
+There were two fundamental ideas in the code we just wrote, which correspond to
+the monad operations of `return` and `bind`.
+
+The first was upgrading a value from `int` to `int * string` by pairing it with
+the empty string. That's what `e` does. We could rename it `return`:
+
+```ocaml
+let return (x : int) : int * string = (x, "")
+```
+This function has the *trivial effect* of putting a value into the metaphorical
+box along with the empty log message.
+
+The second idea was factoring out code to handle pattern matching against pairs
+and string concatenation. Here's that idea expressed as its own function:
+
+```ocaml
+let ( >>= ) (m : int * string) (f : int -> int * string) : int * string =
+  let (x, s1) = m in
+  let (y, s2) = f x in
+  (y, s1 ^ s2)
+```
+
+Using `>>=`, we can re-implement `loggable`, such that no pairs
+or pattern matching are ever used in its body:
+
+```ocaml
+let loggable (name : string) (f : int -> int) : int * string -> int * string =
+  fun m ->
+    m >>= fun x ->
+    log name f x
+```
+
+**The Writer Monad.** The monad we just discovered is usually called the *writer
+monad* (as in, "additionally writing to a log or string"). Here's an
+implementation of the monad signature for it:
+
+```ocaml
+module Writer : Monad = struct
+  type 'a t = 'a * string
+
+  let return x = (x, "")
+
+  let ( >>= ) m f =
+    let (x, s1) = m in
+    let (y, s2) = f x in
+    (y, s1 ^ s2)
+end
+```
+
+As we saw with the maybe monad, these are the same implementations of `return`
+and `>>=` as we invented above, but without the type annotations to force them
+to work only on integers. Indeed, we never needed those annotations; they just
+helped make the code above a little clearer.
+
+It's debatable which version of `loggable` is easier to read. Certainly you need
+to be comfortable with the monadic style of programming to appreciate the
+version that uses `>>=`. But if you were developing a much larger code
+base (i.e., with more functions involving paired strings than just `loggable`),
+using the `>>=` operator is likely to be a good choice. It means the code you
+write can concentrate on the `'a` in the type `'a Writer.t` instead of on the
+strings. In other words, the writer monad will take care of the strings for you,
+as long as you use `return` and `>>=`.
+
+## Example: The Lwt Monad
+
+By now, it's probably obvious that the Lwt promises library that we discussed is
+also a monad. The type `'a Lwt.t` of promises has a `return` and `bind`
+operation of the right types to be a monad:
```

Comment:
```suggestion
By now, it's probably obvious that the Lwt `promises` library that we discussed is
also a monad. The type `'a Lwt.t` of `promises` has a `return` and `bind`
operation of the right types to be a monad:
```

This way if `promises` is the name of a library, which is how it reads (albeit a little confusing). Or does the Lwt library promise something?